src/libs/histogram.c

```diff
@@ -208,36 +211,37 @@ static void _lib_histogram_process_waveform(dt_lib_histogram_t *d, const float *
 
   // Every bin_width x height portion of the image is being described
   // in a 1 pixel x wf_height portion of the histogram.
+  // NOTE: if constant is decreased, will brighten output
   const float brightness = wf_height / 40.0f;
   const float scale = brightness / (height * bin_width);
 
   // 1.0 is at 8/9 of the height!
   const float _height = (float)(wf_height - 1);
 
+  const size_t in_stride = 4U * width;
```

Comment:
probably `(size_t)4U * width` needed here. Same below.

Reply:
Only if the image width is greater than half a billion pixels :-).

Reply:
Yes, I've been just mulling this! And some of the `size_t` I already wrote is a bit silly, as the output buffer is generally about 360x175 pixels! Code such as b43f62be6c56ed80fc746e109f1672f92ee8e131 looks like a good example, as its legible and should do the right thing.

I understand that so long as the array offsets are `size_t`, it may be good for everything which builds them to be size_t. But then so many more questions. Do I write `for(size_t in_y = 0U; in_y < height; in_y++)` with `0U` because I don't want to risk that a compiler would run-time convert `0` to `0U`? But how do we know `U` is the size of of a `size_t`?

And what if I write `const size_t out_y = isnan(v) ? 0U : CLAMPS((size_t) (v * _height), 0U, _height);` and `v` is negative? What happens when casting a negative number to `size_t` before clamping it... Going down the wormhole of unsigned types.

Reply:
For the record, adding these test lines to the function:

```c
  const int a = 175;
  printf("%zu %zu\n", (size_t)(-1.0f * 175), (size_t)(-1.0f * a));
  printf("%d %zu\n", d->waveform_height, (size_t)(-1.0f * wf_height));
  printf("%zu %zu\n", CLAMPS((size_t)(-1.0f * wf_height), 0U, 175U), (size_t)CLAMPS((-1.0f * wf_height), 0.0f, 175.0f));
```

produces


```
0 0
175 18446744073709551441
175 0
```

I'll think out all these types in a reasonable way and make a PR.

Reply:
LOL

generally `(smaller signed type) * (bigger unsigned type)` makes the multiplication work in `smaller signed type` causing potential problems. In this case it's ` (unsigned int) * (signed int)` making whole multiplication work in larger type still...

And floats/doubles are evil too:  `float * int` makes multiplication in float (afaik) but `int *float` and whole multiplication is in int... bah ;) complex stuff. Generally one needs to make sure that multiplication is always done in biggest type unless overflows are intended (and they very hardly ever are) ;)

Reply:
I appreciate the explanation!

Questions, so long as we're doing some explaining, and I don't want to just blindly copy code:
1. Should a loop variable always be `size_t` if it's being used to index an array (or offset a pointer)?
2. Why not `ptrdiff_t`, offsetting pointers is like a pointer difference?
3. What if we know the loop variable will be <= INT_MAX? Should it still be `size_t`? Or `int`, as that's the friendliest type for the processor?
4. What if the loop variable is `int` and it is multiplied by a constant to create an offset. Should it definitely be `size_t` then?
5. If it is multiplied by 4, should that always be written as `4U`?
6. If so, is this for compiler optimization or to make the multiplication work as unsigned?
7. What if it is multiplied by or added to other variables to produce the pointer offset. Should the other variables always be `size_t`?
8. If so, is that to make sure that all the math works in the larger type, or for the sake of optimization, so that they aren't converted to the larger type on the fly?
9. If we know that all the math for the pointer offset will produce a number <= INT_MAX, should all the variables to produce the offset be `int` as that's the friendliest type for the processor?
10. What if the constant is `0`, should it still be written `0U`?

Am going too deep into this, I know...

Reply:
Remember: I'm a certified dumbass, whatever I say might be false and/or stupid. Everything i say comes straight from my tiny brain... that being said:

1 & 2. "depends". Usually if loop index is guaranteed to never be below zero and is not used for pointer arithmetics, then `size_t` is perfectly fine. If loop index will be below zero and used for pointer arithmetics then `ptrdiff_t` is recommended.
3. "depends" too - on 64-bit system `int` might be 32 bit while `size_t` will be 64 bit and usually full bit is better.
4 & 5. if index is `int` then simply `int * 4` is fine, if it's `size_t`, then `size_t * 4U` would probably make it more obvious
6. it's more to prevent unwanted compiler behaviour. Usually you don't want multiplication done in smaller type if you know sizes might overflow it. So generally it's better to do multiplication in bigger type
7. "depends". things producing negatives should be in signed type, but if resulting variable shouldn't be negative then variable should be unsigned... The singed/unsigned & type size business bites :confused: 
8. Generally to make sure math is done in largest possible type without overflows.
9. still - depends, int might not be friendlest for processor
10. hmm... `0` is `int`, `0U` is `unsigned`, `0.f` is float, `0.` is double... but 0*whatever = 0 so it doesn't matter :P (except float/double where that's some magic)

I think clang12 will have warns about multiplication type undersizing.

Reply:
@johnny-bit, both of your statements are false: https://en.cppreference.com/w/c/language/conversion

Specifically:

- `(smaller signed type) * (bigger unsigned type)` is done in bigger unsigned (or at least int) type
- both `int * float` and `float * int` are done in float

For usual arithmetic operations order of 2 operands doesn't matter for resulting type, but for `a * b * c` you have to remember that first `a * b` will be computed and type of `c` doesn't affect that calculation. So type needs to be manually upgraded if result might not fit.

Reply:
Regarding how to write numeric constants (4, 4u, 4.0, 4.0f) - I think it doesn't matter almost any time due to how implicit conversions work, with the exception of using 4.0 in float calculations (which will force it to be done in double). Unless you of course do something intentionally stupid/evil like writing 4.0 in integer calculation.

Reply:
Thanks for all this insight. I pushed another version which makes (I hope) as few changes as possible, and ideally doesn't break MacOS again.